files/ja.7
.\" -*- nroff -*-
.
.TH JA 7
.
.\" ------------------------------------------------------------------
.SH NAME
.\" ------------------------------------------------------------------
.
ja \- enhance Japanese text processing.
.
.ds CW-font CR
.de CW
.  if \\n[.$] \{\
.    rm cw-result
.    if (\\n[.$] >= 3) .as cw-result \\$3\"
.    ie '\\*[\\$0-font]'' .ds cw-font \\$0
.    el .ds cw-font \\*[\\$0-font]
.    as cw-result \f[\\*[cw-font]]\\$1\f[R]\"
.    if (\\n[.$] >= 2) .as cw-result \\$2\"
.    nop \&\\*[cw-result]
.    rm cw-result
.    rm cw-font
.  \}
..
.als HR CW
.
.\" ------------------------------------------------------------------
.SH SYNOPSYS
.\" ------------------------------------------------------------------
.
.EX
$ man -Tpdf -Lja groff.7 >a.pdf
$ zcat `man -w -Lja groff.7` | groff -Tpdf -kt -mja -mandoc - >a.pdf
$ perldocjp -u \c
.I podname \c
.gcolor grey50
| perl -lpe "s/\e"(\e\ew<.*?>)\e"/\e`\e`\e$1''/g" \c
.gcolor
\e
| pod2man --utf8 --fixed=CR \c
| groff -k -Tpdf -mandoc -mja - >a.pdf
.EE
.
pod2man の出力で verbatim 段落の行間を狭くするには、
.EX
| sed -e '/^\e.TH/i.am Vb\en.setlocale english\en..\en.am Ve\en.setlocale\en..'
.EE
.
.
groff を作り直すには、
.EX
$ git clone https://git.savannah.gnu.org/git/groff.git && cd groff
.\" $ git reset --hard
.\" $ git checkout master
.\" $ git pull origin master
$ git checkout 1.24.0
$ patch -p1 <troff-ja.diff
$ ./bootstrap
$ ./configure --prefix=$HOME
$ make
.EE
.
.\" ------------------------------------------------------------------
.SH DESCRIPTION
.\" ------------------------------------------------------------------
.
groff 1.24.0 (troff) の日本語処理に以下の機能を追加します。
.
.PD 0
.IP (1) 4
日本語の行末の改行やマクロが付加する改行を削除します。
(環境を切り替えるものは除きます。)
.
.IP (2)
プロポーショナルフォント (opentype palt など) で、
全角の幅の約物から取り除かれるアキの代りにワードスペースを配置します。
字間にも0幅スペースを配置します。
(スペースは
.CW \e&
で抑止できます。)
.
.IP (3)
日本語処理の動作は
.CW .ja
リクエストで指定します。
.
.IP (4)
ja.tmac を\c
.UR https://www.w3.org/TR/jlreq/
日本語組版処理の要件 (JLreq)
.UE
に近付けます。
字体に含まれるアキの配置も定義します。
.
.IP (5)
1.24.0 の break_char_node は、カーニングやイタリック補正 (\e/) ができ
ないという制限があります。
そのうち直るかもしれませんが、上の (1) (2) を妨げるため、
ここで修正します。
.
.PD
.
.PP
修正した troff では、
.UR https://qiita.com/CodeOne/items/43d2b8e4247b020652b2
日本語文章中、英単語の両端にスペースをつける人
.UE
のように入力する必要がありますが、
.
表の項目に日本語が含まれるときでも
.CW T{
と
.CW T}
で囲む必要がなくなります。
.
.
.\" ------------------------------------------------------------------
.SS ja.tmac
.\" ------------------------------------------------------------------
.
ja.tmac は日本語の文字のフラグを定義します。
groff 1.24.0 で定義されているものと合わせて示します。
.
.RS +2
.TS
tab(;);
l l l.
.ft HR
フラグ;内部名;説明
.ft
.vs -2
_
.T&
n l l.
128;PROHIBITS_BREAK_BEFORE;T{
.UR https://www.w3.org/TR/jlreq/#cl-02
cl-02
.UE
類。直前のブレークを禁止する。
T}
256;PROHIBITS_BREAK_AFTER;T{
.UR https://www.w3.org/TR/jlreq/#cl-0
cl-01
.UE
類。直後のブレークを禁止する。
T}
512;IS_INTERWORD_SPACE;日本語の文字。前後でブレークを許可する。
1024;HAS_SPACE_AFTER;グリフの後にアキがある。(追加)
2048;HAS_SPACE_BEFORE;グリフの前にアキがある。(追加)
_
.TE
.vs
.RE
.PP
.ne 3
ja.tmac の妥当性は、スクリプト ja.tmac.pl で、
フォント (.otf) と特性 (-F palt) を指定して確認します。
.\" 例を示します。
.
.EX
$ perl ja.tmac.pl -F palt HaranoAjiMincho-Regular.otf
$ perl ja.tmac.pl -F palt NotoSerifCJKjp-Regular.otf
.EE
.
HAS_SPACE_BEFORE (AFTER) を
opentype-feature palt の xadvance と xplacement で確かめます
（横書のみ）。
.
PROHIBITS_BREAK_BEFORE (AFTER) の方は
.UR https://metacpan.org/pod/Unicode::UCD
Unicode::UCD
.UE
でカテゴリや名称を確かめます。
.
.ig
アキのある字体は、字体の前または後に二分アキ、
前後両方に四分アキを配置して全角になるようにデザインされていますが、
この修正では少し小さなワードスペースを使います。
..
.
.\" ------------------------------------------------------------------
.SS .ja リクエスト
.\" ------------------------------------------------------------------
.
この修正で追加される groff の日本語処理の動作を指定するために使用します。
.PD 0
.EX
\&.ja [\c
.I f\c
]\c
	\e\(dq 現在の設定値値は\|\c
.CW \en[.ja] \c
\&\|で得られます。
.EE
.
.RS +2
.TS
tab(;);
rb lb.
フラグ \f(CIf\fP;説明
_
.T&
n l.
1;日本語処理を有効にします。
2;字間に0幅スペースを配置しない。
128;デバグ用に内部情報を出力する。
256;デバグ用に内部情報（詳細）を出力する。
_
.TE
.RE
.PD
.
.\" ------------------------------------------------------------------
.SS 連続する約物
.\" ------------------------------------------------------------------
.
.ds 1) \[u2460]
.ds 2) \[u2461]
.ds 3) \[u2462]
.ds 4) \[u2463]
.ds 5) \[u2464]
.ds 6) \[u2465]
.ds 7) \[u2466]
.ds 8) \[u2467]
.ds 9) \[u2468]
.
.PP
約物の前後にワードスペースが適切に配置されることを
日本語組版処理の要件（日本語版）の「\c
.UR https://www.w3.org/TR/jlreq/#positioning_of_consecutive_opening_brackets_closing_brackets_comma_full_stops_and_middle_dots
3.1.4 始め括弧類，終わり括弧類，読点類，句点類及び中点類が連続する場合の配置方法
.UE 」の\c
.UR https://www.w3.org/TR/jlreq/#fig2_13
図69
.UE \c
で確かめます。
.
.de MC
.  nr mc*ll \\n[.l]
.  nr mc*po \\n[.o]
.  nr mc*in \\n[.i]
.  ie (\\n[.$]>=1) .nr mc*cw \\$1
.  el              .nr mc*cw 44n
.  ie (\\n[.$]>=2) .nr mc*gw \\$2
.  el              .nr mc*gw 1n
.  1C
..
.de RC
.  br
.  mk mc*here
.  nr mc*there -\\n[mc*here]u
.\" .  nr mc*there -1v
.  if (\\n[mc*there] > 0) .sp \\n[mc*there]u
.  in \\n[mc*in]u
.  po \\n[mc*po]u
.  ll \\n[mc*ll]u
..
.de 1C
.  RC
.  mk mc*there
.  mk
.  in 0
.  po (\\n[mc*po]u + \\n[mc*in]u)u
.  ll \\n[mc*cw]u
..
.de 2C
.  br
.  mk mc*here
.  if (\\n[mc*here] > \\n[mc*there]) .nr mc*there \\n[mc*here]
.  rt
.  in 0
.  po (\\n[mc*po]u + \\n[mc*in]u)u
.  po +(\\n[mc*cw]u + \\n[mc*gw]u)u
.  ll \\n[mc*cw]u
..
.de 3C
.  2C
.  po +(\\n[mc*cw]u + \\n[mc*gw]u)u
..
.PP
.MC 27n 2n
.HR スペースあり
.nr ss.bak \n[.ss]
.ss (\n[.ss] * 2)
.nf
\*(1) …である。」この…
\*(2) …である）。この…
\*(3) …である、「この…
\*(4) …である」「この…
\*(5) …である「『この…
\*(6) …である）」この…
\*(7) …編集」・「校正…
.fi
.ss \n[ss.bak]
.
.2C
.HR "スペースなし (\e& で抑止)"
.nf
\*(1) …である。」\&この…
\*(2) …である）。\&この…
\*(3) …である、\&「この…
\*(4) …である」\&「この…
\*(5) …である\&「『この…
\*(6) …である）」\&この…
\*(7) …編集」・\&「校正…
.fi
.
.3C
.HR "スペースなし (等幅フォント)"
.ft CR
.ps +2
.nf
\*(1) …である。」この…
\*(2) …である）。この…
\*(3) …である、「この…
\*(4) …である」「この…
\*(5) …である「『この…
\*(6) …である）」この…
\*(7) …編集」・「校正…
.fi
.ft
.ps
.RC


files/ja.tmac.pl
#!/usr/bin/env perl

use strict;
use warnings;

use utf8;
use Encode;
use Getopt::Long qw(:config gnu_getopt);

my %ot_feature;
GetOptions(
    "w=i" => \ my $space_width,
    "F=s" => sub { my ($k, $v) = split '=', $_[1]; $v //= "*,*"; $ot_feature{$k} = $v },
    "I=s" => \ my @include_path,
    "L=s" => \ my $lang,
    "v" => \ my $verbose,
    "D" => \ my $debug,
);


$lang //= 'ja';

use Unicode::UCD qw/charinfo/;
use Font::TTF::Font;

use feature 'say';
use Test::More;
#use Data::Dumper qw/Dumper/;
#$Data::Dumper::Indent = 1;
#$Data::Dumper::Terse = 1;

my ($prog) = $0 =~ /[^\/](.*)/;
$prog ||= $0;
my $otffile = shift;
unless ($otffile) {
    warn "prog: no otffile\n";
    usage();
}

sub usage {
    die "usage: $prog [-w space-width] [-F opentype-feature][-I tmac_path ] otffile\n";
}

my $otf = Font::TTF::Font->open($otffile);
unless ($otf) {
    die "can't open $otffile\n";
}

my @gsub_index;
my @gpos_index;

while (my ($k, $v) = each %ot_feature) {
    next if $k =~ /liga|kern/;
    my $option = join ',', ((split /,/, $v // "*,*"), $k)[0..2];
    push @gsub_index, grep defined, ot_feature($otf, 'GSUB', $option);
    push @gpos_index, grep defined, ot_feature($otf, 'GPOS', $option);
}

my $gsub = gsub($otf, @gsub_index);
my $gpos = gpos($otf, @gpos_index);

$otf->{'CFF '}->read;
my $gid2cid = $otf->{'CFF '}->Charset->{code};

$otf->{'hmtx'}->read;
$otf->{'cmap'}->read;

use constant {
    PROHIBITS_BREAK_BEFORE => 0x080,
    PROHIBITS_BREAK_AFTER  => 0x100,
    IS_INTERWORD_SPACE     => 0x200,
    HAS_SPACE_AFTER        => 0x400,
    HAS_SPACE_BEFORE       => 0x800,
};

my %re = (
    'cl-03' => qr/Wave Dash|Hyphen|En Dash/i, # $charinfo->{category} =~ /Pd/
    'cl-04' => qr/Exclamation Mark|Question Mark/i,
    'cl-05' => qr/Middle Dot|colon/i,
    'cl-06' => qr/Full Stop/i,
    'cl-07' => qr/Comma/i,
    'cl-08' => qr/Em Dash|Two Dot Leader|Horizontal Ellipsis|Vertical Kana Repeat/i,
    'cl-09' => qr/Ditto Mark|Iteration Mark/i,
    'cl-10' => qr/Prolonged Sound Mark/i,
    'cl-11' => qr/(?:Hiragana|Katakana) Letter Small/i,
);

sub get_flags {
    my $flags = 0;

    my $uv = shift;
    local $_ = pack "U*", $uv;

    return undef unless my $charinfo = charinfo($uv);
    if ($charinfo->{category} =~ /P[si]/) {
	$flags |= PROHIBITS_BREAK_AFTER;
    }
    if ($charinfo->{category} =~ /P[efd]/) {
	$flags |= PROHIBITS_BREAK_BEFORE;
    }
    if ($charinfo->{name} =~ /$re{'cl-04'}|$re{'cl-05'}|$re{'cl-06'}
			     |$re{'cl-07'}|$re{'cl-08'}|$re{'cl-09'}
			     |$re{'cl-10'}|$re{'cl-11'}/x) {
	$flags |= PROHIBITS_BREAK_BEFORE;
    }
    if ($charinfo->{block} =~ /Hiragana|Katakana|CJK|Halfwidth and Fullwidth Forms/i) {
	$flags |= IS_INTERWORD_SPACE;
    }
    if ((!$flags || $charinfo->{name} =~ /xxxxx/) && $debug) {
	printf "U+%04X = 0x%03x\n", $uv, $flags;
	printf "U+%04X = %s\n", $uv, join ' ', explain $charinfo;
    }

    if (my $gid = $otf->{'cmap'}->find_ms->{val}{$uv}) {
	if (my $subst = find_gsub($gid)) {
	    $gid = $subst;
	}

	my %x;
	for ($otf->{'hmtx'}{'advance'}[$gid]) {
	    $x{advance} = $_ if defined;
	}

	if ($gpos) {
	    for my $k (keys %{$gpos->{$gid}}) {
		for ($gpos->{$gid}{$k}) {
		    $x{lc $k} = $_ if defined;
		}
	    }
	}

	if (defined $x{advance} && defined $x{xadvance} && !defined $x{xplacement}) {
	    if (near_eq($x{advance}, 1000) &&
		near_eq($x{xadvance}, -500)) {
		$flags |= HAS_SPACE_AFTER;
	    }
	} elsif (defined $x{advance} && defined $x{xadvance} && defined $x{xplacement}) {
	    if (near_eq($x{advance}, 1000) &&
		near_eq($x{xadvance}, -500) &&
		near_eq($x{xplacement}, -500)) {
		$flags |= HAS_SPACE_BEFORE;
	    } elsif (near_eq($x{advance}, 1000) &&
		     near_eq($x{xadvance}, -500) &&
		     near_eq($x{xplacement}, 0)) {
		$flags |= HAS_SPACE_AFTER;
	    } elsif (near_eq($x{advance}, 1000) &&
		     near_eq($x{xadvance}, -500) &&
		     near_eq($x{xplacement}, -250)) {
		$flags |= HAS_SPACE_AFTER | HAS_SPACE_BEFORE;
	    }
	}
    }
    $flags;
}


sub near_eq {
    my ($a, $b, $e) = @_;
    $e //= 0.03;
    my $d = $b - $a;
    my $eq = ($d / 1000) >= -$e && ($d / 1000) <= +$e;
    #printf STDERR "$a - $b = $d (%g) = %s\n", $d, $eq ? 'eq' : 'neq';
    $eq;
}


#my %gid_hint;

sub find_gsub {
    my ($gid, $hint) = @_;
    if ($gsub) {
	my $start = $gid;
	while (my $subst = $gsub->{$gid}) {
	    if ($subst == $start) {
		warn "gsub $start seems looping." if $debug;
		last;
	    }
	    $gid = $subst;
	}
	if ($gid != $start) {
	    #$gid_hint{$gid} = $hint if $hint;
	    return $gid;
	}
    }
    undef;
}


sub gsub {
    my $otf = shift;
    my $gsub;
    for my $index (grep defined, @_) {
	my $value = $otf->{GSUB}{LOOKUP}[$index];
	if ($value->{TYPE} == 1) {
	    for (@{$value->{SUB}}) {
		while (my ($gid, $i) = each %{$_->{COVERAGE}{val}}) {
		    $gsub->{$gid} = $_->{RULES}[$i][0]{ACTION}[0];
		}
	    }
	} elsif ($value->{TYPE} == 4) {
	    for (@{$value->{SUB}}) {
		while (my ($gid, $i) = each %{$_->{COVERAGE}{val}}) {
		    for (@{$_->{RULES}[$i]}) {
			$gsub->{join $;, @{$_->{ACTION}}} =
			    [ $gid + 0, @{$_->{MATCH}} ];
		    }
		}
	    }
	} else {
	    die "gsub: unknown \$value->{TYPE}: $value->{TYPE}";
	}
    }
    $gsub;
}


sub gpos {
    my $otf = shift;
    my $gpos;
    for my $index (grep defined, @_) {
	my $value = $otf->{GPOS}{LOOKUP}[$index];
	if ($value->{TYPE} == 1) {
	    # Lookup type 1 subtable: single adjustment positioning
	    for (@{$value->{SUB}}) {
		while (my ($gid, $i) = each %{$_->{COVERAGE}{val}}) {
		    $gpos->{$gid} = $_->{RULES}[$i][0]{ACTION}[0];
		}
	    }
	} elsif ($value->{TYPE} == 2) {
	    # Lookup type 2 subtable: pair adjustment positioning
	    for (@{$value->{SUB}}) {
		my @gid;
		while (my ($gid, $i) = each %{$_->{COVERAGE}{val}}) {
		    $gid[$i] = $gid;
		}
		my $MATCH_TYPE  = $_->{MATCH_TYPE};
		my $ACTION_TYPE = $_->{ACTION_TYPE};
		if ($MATCH_TYPE eq 'g' && $ACTION_TYPE eq 'p') {
		    # $_->{FORMAT} = 1: Pair adjustment positioning
		    # MATCH_TYPE = 'g': A glyph array
		    # ACTION_TYPE = 'p': Pair adjustment
		    for my $i (0 ..  $#{$_->{RULES}}) {
			my $gid = $gid[$i];
			for my $j (0 ..  $#{$_->{RULES}[$i]}) {
			    my $gid2 = $_->{RULES}[$i][$j]{MATCH}[0];
			    $gpos->{$gid, $gid2} = $_->{RULES}[$i][$j]{ACTION}[0];
			}
		    }
		} elsif ($MATCH_TYPE eq 'c' && $ACTION_TYPE eq 'p') {
		    # $_->{FORMAT} = 2: Pair adjustment positioning
		    # MATCH_TYPE = 'c': An array of class values
		    # ACTION_TYPE = 'p': Pair adjustment
		    for my $gid (@gid) {
			my $c = $_->{CLASS}{val}{$gid};
			next unless defined $c;
			while (my ($gid2, $c2) = each %{$_->{MATCH}[0]{val}}) {
			    next unless $c2;
			    $gpos->{$gid, $gid2} = $_->{RULES}[$c][$c2]{ACTION}[0];
			}
		    }
		} else {
		    die "gpos: unknown \$_->{FORMAT}: $_->{FORMAT} in TYPE 2";
		}
	    }
	}
	else {
	    die "gpos: unknown \$value->{TYPE}: $value->{TYPE} (index: $index)";
	}
    }
    $gpos;
}


sub ot_feature {
    my ($otf, $tag, $option) = @_;
    return undef unless $otf;
    return undef unless $option;

    my ($script, $lang, $wanted) = split /,\s*/, $option;
    return undef unless defined $script && defined $lang && defined $wanted;

    unless (ref $otf->{$tag} && $otf->{$tag}->read) {
	warn "$prog: can't read $tag table; ignored\n";
	return undef;
    }

    # filter /$script/,
    my %seen;
    my @script = grep !$seen{$_}++,
	grep $script eq '*' || /^$script\s*$/,
	keys %{$otf->{$tag}{SCRIPTS}};

    # filter /lang/, @script
    my @lang;
    my @features = map {
	my $languages = $otf->{$tag}{SCRIPTS}{$_};
	@lang = grep !$seen{$_}++, @{$languages->{LANG_TAGS}};
	@lang = 'DFLT' unless @lang;
	map @{$_->{FEATURES}},
	    map $languages->{$_} || $languages->{DEFAULT},
	    grep $lang eq '*' || /^$lang\s*$/i,
	    @lang;
    } @script;

    # filter  /wanted/, @features
    %seen = ();
    my @index = grep !$seen{$_}++,
	map @{ $otf->{$tag}{FEATURES}{$_}{LOOKUPS} },
	grep /$wanted/,
	@features;

    1 and printf STDERR "$tag: script => '%s', lang => '%s', features => '%s': %s\n",
	$script, $lang, $wanted, join ', ', @index
	if @index;

    @index;
}


my %cflags;

sub parse_lang_tmac {
    my $tmac = shift // "$lang.tmac" ;
    unless (@include_path) {
	push @include_path, grep defined,
	    $ENV{GROFF_TMAC_PATH},
	    map { "$_/site-tmac:$_/current/tmac" }
	    grep -d, "$ENV{HOME}/share/groff", "/usr/local/share/groff", "/usr/share/groff";
    }
    my $include_path = join ':', @include_path;
    open my $fd, "-|", qq[echo '.so $tmac' |soelim -I $include_path];
    local $_ = join '', <$fd>;
    close $fd;
    s/\\".*//gm;
    s/\\(#.*)?\n//gm;
    for (grep /^.\s*cflags/, split "\n") {
	if (s/^[.]\s*cflags\s+(\d+)\s+//) {
	    my $n = $1;
	    my @class;
	    my @char;
	    while (length > 0) {
		if (s/\\C["'](.*?)["']//) {
		    push @class, $1; # class
		} elsif (s/\\\[u(.*?)\]-\\\[u(.*?)\]//) {
		    push @char, map { sprintf "u%04X", $_ } hex($1) .. hex($2);
		} elsif (s/\\\[(u.*?)\]//) {
		    push @char, $1;
		} elsif (s/\\\[(.*?)[\]]// || s/\\\((..)// || s/(\S)//) {
		    #push @char, $1;
		    warn "# .cflags $n $& ignored\n";
		}
		s/\s*//;
	    }
	    $cflags{$n} = [ (map "\\C'$_'", @class), (map "\\[$_]", @char) ];
	}
    }
}


my $error = 0;

sub test_cflags {
    my $n = shift;
    for my $c (@{$cflags{$n}}) {
	say "# $c:";
 	if (my $p = run_pchar($c)) {
	    for (@{$p->{code_points}}) {
		my $hex = sprintf "%04X", $_;
    		if (defined (my $flags = get_flags($_))) {
		    my $ok = $p->{flags} & $flags;
		    $error++ unless $ok;
		    if ($verbose || !$ok) {
			say sprintf "u$hex 0x%03x, 0x%03x - %s",
			    $p->{flags}, $flags, $ok ? 'ok' : 'not ok';
		    }
		} else {
		    say "u$hex ignored" if $verbose;
		}
	    }
	}
   }
}


sub run_pchar {
    my $c = shift;
    open my $fd, "-|", qq[echo ".pchar $c" | groff -Tpdf -m$lang 2>&1];
    my %defined_at;
    my @code_points;
    my $flags;
    while (<$fd>) {
	if (s/defined at:\s*//) {
	    while (s/([^:]*):\s*"([^"]*)"(?:,\s+)?//) {
		$defined_at{$1} = $2;
	    }
	} elsif (s/^\s*contains code points:\s*//) {
	    for (split /\s+/) {
		my ($a, $b) = map { s/U\+//; hex; } split '-';
		push @code_points,  $a .. ($b // $a);
	    }
	} elsif (s/^\s*flags:\s*//) {
	    ($flags) = split /\s+/;
	}
    }
    { defined_at => \%defined_at, code_points => \@code_points, flags => $flags };
}

parse_lang_tmac;
test_cflags HAS_SPACE_AFTER;
test_cflags HAS_SPACE_BEFORE;
test_cflags PROHIBITS_BREAK_BEFORE;
test_cflags PROHIBITS_BREAK_AFTER;
test_cflags IS_INTERWORD_SPACE;

exit 1 if $error;
exit 0;

# Local Variables:
# fill-column: 72
# tab-width: 8
# indent-tabs-mode: t
# mode: CPerl
# cperl-indent-level: 4
# cperl-continued-statement-offset: 4
# cperl-close-paren-offset: -4
# cperl-indent-parens-as-block: t
# End:
# vim: set cindent noexpandtab shiftwidth=2 softtabstop=2 textwidth=72:
